src/store/hooks/useAuth.tsx

```diff
@@ -0,0 +1,223 @@
+      determiningStatus: true,
+      isAdmin: true,
+    });
+    setAdminVerification(false);
+    const auth = getAuth();
+    signOut(auth).then(() => {
+      setUser({
+        user: null,
+        determiningStatus: false,
+        isAdmin: false,
+      });
+    }).catch((error) => {
+      setUser({
+        user: null,
+        determiningStatus: false,
+        isAdmin: false,
+      });
+      console.error(`There was an issue signing-out the user: ${error.message}`);
+    });
+  };
+
+  function isFirebaseUser(input: verifyAdminStatusInput): input is User {
+    return input !== null && typeof input === 'object' && 'uid' in input;
+  }
+
+  function isLocalStorageUser(input: verifyAdminStatusInput): input is LocalStorageUser {
+    return input !== null && typeof input === 'object' && !('uid' in input);
+  }
+
+  const verifyAdminStatus = async (inputUser: verifyAdminStatusInput) => {
+    if (inputUser === null) {
+      return Promise.resolve(true);
+    } if (isFirebaseUser(inputUser)) {
+      // Assert that the inputted user is the User type
+      const firebaseUser = inputUser as User;
+      if (storageEngine instanceof FirebaseStorageEngine) {
+        const isAuthEnabled = await storageEngine?.getUserManagementData('authentication').then((data) => data?.isEnabled);
+        if (isAuthEnabled !== false) {
+          // Validate the user if Auth enabled
+          return await storageEngine.validateUserAdminStatus(firebaseUser, globalConfig.adminUsers);
+        }
+        // If authDisabled, allow access
+        return Promise.resolve(true);
+      }
+      // If the user is Firebase but the storageEngine is not, something has gone wrong. Deny access
+      return Promise.resolve(false);
+    } if (isLocalStorageUser(inputUser)) {
+      if (storageEngine) {
+        // If storageEngine defined, allow access
+        return Promise.resolve(true);
+      }
+      // If localStorageUser but no storageEngine defined, deny access.
+      return Promise.resolve(false);
+    }
+    // Some other forms of the user input should be denied access.
+    return Promise.resolve(false);
+  };
+
+  useEffect(() => {
+    // Set initialUser
+    setUser({
+      user: null,
+      isAdmin: false,
+      determiningStatus: true,
+    });
+
+    // Get authentication
+    let auth: Auth;
+    try {
+      auth = getAuth();
+    } catch (error) {
+      console.warn('No firebase store.');
+    }
+
+    // Handle auth state changes for Firebase
+    const handleAuthStateChanged = async (firebaseUser: User | null) => {
+      // Reset the user
+      setUser({
+        user: null,
+        isAdmin: false,
+        determiningStatus: true,
+      });
+      if (firebaseUser) {
+        // Reach out to firebase to validate user
+        const isAdmin = await verifyAdminStatus(firebaseUser);
+        setAdminVerification(true);
+        login({
+          user: firebaseUser,
+          determiningStatus: false,
+          isAdmin,
+        });
```

Comment:
Setting the user async here is making me a little nervous

Reply:
This is meant to be the callback for the onAuthStatusChanged function. So I don't think the asynchronous behavior is an issue -- it's a listener which, whenever the auth status changes, the user gets reverified. I am going to remove the initial reset of the user though. We don't want the user to get reset if they are indeed an admin and their refresh token just expired. Instead, this'll be changed to just change the determining status of the user.